openedx/openedx-platform
# fix: delete discarded components from meilisearch index [FC-0062]

## Description

Reverting a library to its last published state is done by calling [revert_changes](https://github.com/open-craft/edx-platform/blob/34d2c0ec579697faa4111a587dbe3042b39cdbb4/openedx/core/djangoapps/content_libraries/api.py#L986) view, which triggers `CONTENT_LIBRARY_UPDATED` signal. This PR updates the [task](https://github.com/open-craft/edx-platform/blob/34d2c0ec579697faa4111a587dbe3042b39cdbb4/openedx/core/djangoapps/content/search/tasks.py#L75) that is triggered to also delete newly created components which were never published from the meilisearch index.

This PR includes three ways of updating the index:

* 970736fbd6f3f64e998a3db669a1ecfdaeb4f1b9 & 06e4c5d1af12cc80e8db336d48f277ca4aa3b074 fetches draft components that have version = null and explicitly removes them the index. One drawback here is that if the number of components having draft version = null increase, this technique will try to delete them on each discard/publish even if they don't exists in the index.
* 30ea1897f9e691080beae30b645b7206035c5c84 deletes all documents for a given library in the index. This works but it also deletes components which do have draft version and adds them back.
* 2ec9070c726b9b122a75b9c868d76490520699b9 only deletes components that were never published after the publish or discard request.

## Supporting information

* https://github.com/openedx/frontend-app-course-authoring/issues/1200
* ~https://github.com/open-craft/frontend-app-course-authoring/pull/58~ https://github.com/openedx/frontend-app-course-authoring/pull/1214
* https://github.com/openedx/openedx-learning/pull/207

## Testing instructions

See https://github.com/open-craft/frontend-app-course-authoring/pull/58 for instructions on testing.

## Deadline

"None" if there's no rush, or provide a specific date or event (and reason) if there is one.

## Other information

Include anything else that will help reviewers and consumers understand the change.

- Does this change depend on other changes elsewhere?
- Any special concerns or limitations? For example: deprecations, migrations, security, or accessibility.
- If your [database migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) can't be rolled back easily.
